driver-core/src/main/com/mongodb/internal/operation/AsyncSingleBatchCursor.java

```diff
@@ -19,19 +19,26 @@
 import com.mongodb.MongoException;
 import com.mongodb.internal.async.AsyncBatchCursor;
 import com.mongodb.internal.async.SingleResultCallback;
-import com.mongodb.internal.connection.QueryResult;
 
 import java.util.List;
 
-import static com.mongodb.assertions.Assertions.isTrue;
+import static java.util.Collections.emptyList;
 
-class AsyncSingleBatchQueryCursor<T> implements AsyncBatchCursor<T> {
-    private volatile QueryResult<T> firstBatch;
-    private volatile boolean closed;
+class AsyncSingleBatchCursor<T> implements AsyncBatchCursor<T> {
 
-    AsyncSingleBatchQueryCursor(final QueryResult<T> firstBatch) {
-        this.firstBatch = firstBatch;
-        isTrue("Empty Cursor", firstBatch.getCursor() == null);
+    static <R> AsyncSingleBatchCursor<R> createEmptyAsyncSingleBatchCursor(final int batchSize) {
+        return new AsyncSingleBatchCursor<>(emptyList(), batchSize);
+    }
+
+    private final List<T> batch;
+    private final int batchSize;
+
+    private volatile boolean hasNext = true;
```

Comment:
Valentin has commented previously that volatile variables like this are unnecessary.

Reply:
Previously queryresult was volatile and that has been replaced by hasNext.

Not sure where previously it was mentioned but usage wasn't changed in AsyncSingleBatchQueryCursor, so I kept the same behaviour